File: apps/desktop/src/lib/control/selectedTaskReviewOutcomeRoute.ts
import { parseSingleRecordResponse, type QueryFallback } from "./singleRecordResponse";
import type { ControlResponseEnvelopeDto } from "./envelopes";

export type ControlSelectedTaskReviewOutcomeRouteSourceCountsDto = {
  decision_records: number;
  work_item_refs: number;
  evidence_refs: number;
  review_gap_count: number;
  scm_handoff_refs: number;
  downstream_command_hints: number;
  blockers: number;
};

export type ControlSelectedTaskReviewOutcomeRouteNoEffectsDto = {
  review_mutation_performed: boolean;
  task_lifecycle_mutation_performed: boolean;
  provider_execution_performed: boolean;
  provider_write_performed: boolean;
  scm_or_forge_mutation_performed: boolean;
  accepted_memory_apply_performed: boolean;
  planning_apply_performed: boolean;
  projection_write_performed: boolean;
  agent_scheduling_performed: boolean;
  ui_effect_performed: boolean;
};

export type ControlSelectedTaskReviewOutcomeRouteDto = {
  route_id: string;
  project_id: string;
  task_id: string;
  status: string;
  primary_route: string;
  candidates: string[];
  decision_ref: string | null;
  decision_outcome: string | null;
  work_item_refs: string[];
  evidence_refs: string[];
  downstream_command_hints: string[];
  blockers: string[];
  source_counts: ControlSelectedTaskReviewOutcomeRouteSourceCountsDto;
  no_effects: ControlSelectedTaskReviewOutcomeRouteNoEffectsDto;
};

export type SelectedTaskReviewOutcomeRouteQueryResult =
  | {
      state: "record";
      route: ControlSelectedTaskReviewOutcomeRouteDto;
    }
  | QueryFallback;

export function selectedTaskReviewOutcomeRouteFromResponse(
  response: ControlResponseEnvelopeDto,
): SelectedTaskReviewOutcomeRouteQueryResult {
  return parseSingleRecordResponse(response, "selected_task_review_outcome_route", "selected task review outcome route", (body) => ({
    state: "record" as const,
    route: body.route,
  }));
}

File: apps/nucleusd/src/query/tests/selected_task_review_outcome_route.rs
use nucleus_server::{
    ControlSelectedTaskReviewOutcomeRouteDto, ControlSelectedTaskReviewOutcomeRouteNoEffectsDto,
    ControlSelectedTaskReviewOutcomeRouteSourceCountsDto,
};

use super::*;

#[test]
fn selected_task_review_outcome_route_response_lines_are_read_only_and_sanitized() {
    let lines = typed_response::selected_task_review_outcome_route_response_lines(
        "selected-task-review-outcome-route",
        ControlSelectedTaskReviewOutcomeRouteDto {
            route_id: "selected-task-review-outcome-route:task:nucleus-local:bootstrap".to_owned(),
            project_id: "project:nucleus-local".to_owned(),
            task_id: "task:nucleus-local:bootstrap".to_owned(),
            status: "ready".to_owned(),
            primary_route: "ready_for_completion_admission".to_owned(),
            candidates: vec!["ready_for_completion_admission".to_owned()],
            decision_ref: Some(
                "selected-task-review-decision:task:nucleus-local:bootstrap:accept".to_owned(),
            ),
            decision_outcome: Some("accepted".to_owned()),
            work_item_refs: vec!["work:1".to_owned()],
            evidence_refs: vec!["checkpoint:1".to_owned()],
            downstream_command_hints: vec!["complete_selected_task".to_owned()],
            blockers: vec!["downstream_command_not_defined".to_owned()],
            source_counts: ControlSelectedTaskReviewOutcomeRouteSourceCountsDto {
                decision_records: 1,
                work_item_refs: 1,
                evidence_refs: 1,
                review_gap_count: 0,
                scm_handoff_refs: 0,
                downstream_command_hints: 1,
                blockers: 1,
            },
            no_effects: ControlSelectedTaskReviewOutcomeRouteNoEffectsDto {
                review_mutation_performed: false,
                task_lifecycle_mutation_performed: false,
                provider_execution_performed: false,
                provider_write_performed: false,
                scm_or_forge_mutation_performed: false,
                accepted_memory_apply_performed: false,
                planning_apply_performed: false,
                projection_write_performed: false,
                agent_scheduling_performed: false,
                ui_effect_performed: false,
            },
        },
    );
    let rendered = lines.join("\n");

    assert!(rendered.contains("domain=selected-task-review-outcome-route"));
    assert!(rendered.contains("status=ready"));
    assert!(rendered.contains("primary_route=ready_for_completion_admission"));
    assert!(rendered.contains("decision_outcome=accepted"));
    assert!(rendered.contains("task_lifecycle_mutation=false"));
    assert!(rendered.contains("provider_execution=false"));
    assert!(rendered.contains("scm_or_forge_mutation=false"));
    assert!(rendered.contains("client_can_mutate=false"));
    assert!(rendered.contains("payloads_exposed=false"));
    assert!(!rendered.contains("raw_payload"));
    assert!(!rendered.contains("private:context"));
}
